### PR TITLE
Issue/4462 reader traintracks

### LINKED ISSUE #4462: Reader: Add consistent search Tracks events and traintracks events
Reader: Add consistent search Tracks events and traintracks events

Now that search is launched on all three platforms there are a few places where the stats we are tracking do not exactly match each other. Here are the set of Tracks events in Calypso that I *think* we should also implement on mobile (with the appropriate name changes):

- calypso_reader_search_loaded: wordpress.com/read/search is loaded with an empty search string
- calypso_reader_search_performed: a search request has been sent to the API. The web does search as you type with the search getting sent once there is a 500ms delay, but I think there is no reason the mobile apps wouldn't work with a similar event.
- calypso_reader_searchcard_clicked: User clicks on a result to do an article open.
- calypso_traintracks_render (example props: fetch_algo: read:search:posts/1, fetch_blog_id: 123, fetch_post_id:123, query: news): a search result is rendered for the user to see. This means that if the client requests 10 results, then when it gets those results to display (and ideally when it actually shows them to the user) it will fire 10 separate events. See "traintracks" below. 
- calypso_traintracks_interact (example props: action: follow, fetch_algo: read:search:posts/1, fetch_blog_id: 123, fetch_post_id:123, query: news): a search result is interacted with by the user (could also be in the full post view). List of interaction actions is: article_liked, article_opened, searchcard_clicked, site_followed, site_unfollowed. This is sort of a meta event that wraps up other Tracks events in a common way. See "traintrackss below".

Mobile apps also have some support for suggesting searches to users based on their past search history. It would be good to make sure we have tracks events for these and also pass through the query that the user selected. Maybe call it wpios_reader_search_suggest_selected and wpios_reader_search_suggest_shown?

**What are TrainTracks?**

We want a way to improve our algorithms (A/B test new ones) without having to change the clients. TrainTracks is a common method for tracking results from any api endpoint and we are deploying it for related posts, cold start, search, recommendations, etc

Key problem it solves is that a data scientist can develop and deploy an algorithm change and launch it to all platforms with no coordination. This should help is continually improve our algorithms over time. The idea is loosely inspired by this article: http://multithreaded.stitchfix.com/blog/2016/03/16/engineers-shouldnt-write-etl/ Make launching new algorithms routine and streamlined. Avoid having to sync up between the front end and back end. For instance, I ran an A/B test looking at whether to rank more recent content higher last week. And now we are running an A/B test where we let older content rank higher. I don't currently have a way to track these results on our mobile platforms because they aren't sending TrainTracks events.

So what is our framework (currently it may evolve as we find problems):
- When an algorithmic api responds to a request for results we return a "railcar" attached to each result (yes, we're taking this metaphor a little far ;) )
- Railcar contains the following fields:
 - fetch_alg - string identifying the algorithm that generated th
 - fetch_position - integer. The ranked position within the results
 - fetch_lang - the language(s) the algorithm was performed against
 - fetch_query - the user entered query
 - railcar - a unique ID that we use for matching render and interaction events
 - rec_blog_id - the blog id for this result
 - rec_post_id - the post id for this result
 - rec_feed_id - the feed id for this result
 - rec_feed_item_id - the feed item id for this result
 - more to be added in the future
- That's a lot of fields, but as a client of TrainTracks, all the client needs to do is take every property in the railcar and add it as an event property to the render and interact traintracks events
- The API also returns a top level "algorithm" field which is a string that should be used when paging the API. This ensures that we get the same algorithm when paging and so we won't see duplicate results.

**What events trigger "Interaction" events?**

Anything the user can do on result (open, like, comment, follow, etc). There is a list in Calypso here: https://github.com/Automattic/wp-calypso/blob/ca16ae870a6c8df1cb050936f40abb53fc1a2994/client/reader/stats.js#L93

These events then get truncated to remove the prefix and are added to the "action" property: https://github.com/Automattic/wp-calypso/blob/ca16ae870a6c8df1cb050936f40abb53fc1a2994/client/reader/stats.js#L107

cc @xyu

### LINKED ISSUE #5817: Crash report 7.2.1 - "NumberFormatException" in Visual Editor, onMediaRetryClicked
```
Fatal Exception: java.lang.NumberFormatException: Invalid int: "zss_field_content"
       at java.lang.Integer.invalidInt(Integer.java:138)
       at java.lang.Integer.parse(Integer.java:375)
       at java.lang.Integer.parseInt(Integer.java:366)
       at java.lang.Integer.parseInt(Integer.java:332)
       at java.lang.Integer.valueOf(Integer.java:490)
       at org.wordpress.android.ui.posts.EditPostActivity.onMediaRetryClicked(EditPostActivity.java:2219)
       at org.wordpress.android.editor.EditorFragment.onMediaTapped(EditorFragment.java:1378)
       at org.wordpress.android.editor.JsCallbackReceiver.executeCallback(JsCallbackReceiver.java:160)
       at org.wordpress.android.editor.EditorWebViewAbstract$1.shouldOverrideUrlLoading(EditorWebViewAbstract.java:80)
       at android.webkit.CallbackProxy.uiOverrideUrlLoading(CallbackProxy.java:274)
       at android.webkit.CallbackProxy.handleMessage(CallbackProxy.java:376)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:137)
       at android.app.ActivityThread.main(ActivityThread.java:4903)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:511)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:795)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:562)
       at dalvik.system.NativeStart.main(NativeStart.java)

```

### PR DESCRIPTION
**Pull request by nbradbury**
Resolves #4462 - adds initial support for tracking TrainTracks "railcars," which are JSON payloads added to reader posts (currently limited to search results). This payload is then passed when tracking certain events related to reader posts.

There are two new events here: `TRAIN_TRACKS_RENDER` and `TRAIN_TRACKS_INTERACT`. 

`TRAIN_TRACKS_RENDER` records when a reader post is rendered in the list view for the first time for a specific stream.

`TRAIN_TRACKS_INTERACT` records when supported interactions occur on a reader post. There is a list of these in Calypso here: https://github.com/Automattic/wp-calypso/blob/ca16ae870a6c8df1cb050936f40abb53fc1a2994/client/reader/stats.js#L93 , which on Android corresponds to `READER_ARTICLE_LIKED`, `READER_ARTICLE_OPENED`, `READER_SEARCH_RESULT_TAPPED` and `READER_ARTICLE_COMMENTED_ON`. Note that two interactions are missing on Android - issue #4539 has been filed to add them.

To test this, perform a search in the reader, then interact with the search results. Scrolling the list should fire several calls to [trackRailcarRender()](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/4462-reader-traintracks/WordPress/src/main/java/org/wordpress/android/util/AnalyticsUtils.java#L218). Tapping or liking a search result should fire a call to [trackRailcarInteract()](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/4462-reader-traintracks/WordPress/src/main/java/org/wordpress/android/util/AnalyticsUtils.java#L231).

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/util/AnalyticsUtils.java [hunk 1]
```diff
@@ -186,6 +199,72 @@ public static void trackWithReaderPostDetails(AnalyticsTracker.Stat stat, Reader
...
+
+    /*
+     *  Converts the JSON string of a railcar to a properties list using the existing json key names
+     */
+    private static Map<String, Object> railcarJsonToProperties(@NonNull String railcarJson) {
+        Map<String, Object> properties = new HashMap<>();
+        try {
+            JSONObject jsonRailcar = new JSONObject(railcarJson);
+            Iterator<String> iter = jsonRailcar.keys();
+            while (iter.hasNext()) {
```
Comment (roundhill) [inline WordPress/src/main/java/org/wordpress/android/util/AnalyticsUtils.java]: Any performance impact here? I'm assuming it's really lightweight json objects.

Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/util/AnalyticsUtils.java]: Yes, they're fairly lightweight. Here's an example from a post returned by `/v1.2/read/search/?q=android`:

screen shot 2016-09-09 at 6 04 35 pm

**Comment by roundhill** [comment]
Looks good, I'm seeing all of the events fire when they should. One question on `READER_ARTICLE_LIKED`, you can spam it by liking, then unliking, then liking the post again. Not sure if it works that way in the web reader?

Either way go ahead and :shipit:

**Comment by nbradbury** [comment]
Thanks for the review, @roundhill! I don't think the like/unlike/like issue will be a problem since this is used for internal analytics rather than for an author's stats (so there's no real value in spamming via multiple likes).

At any rate, on Android it's up to the reviewer to merge, so feel free to merge when ready :)

**Comment by roundhill** [comment]
> At any rate, on Android it's up to the reviewer to merge, so feel free to merge when ready :)

Oh yeah, how quickly I forgot! :shipit:

**Pull request merged**